File: Leetcode Daily Challenge/027-September-2025/024-fraction-to-recurring-decimal.cpp
// https://leetcode.com/problems/fraction-to-recurring-decimal/




#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    using ll = long long;
    string fractionToDecimal(int numr, int deno) {
        if (numr == 0) return "0";

        string result;
        if ((ll)numr * deno < 0) {
            result += "-";
        } 

        ll absNumr = labs(numr);
        ll absDeno = labs(deno);

        ll div  = absNumr / absDeno;
        result += to_string(div);   // log10(div)
        
        ll rem  = absNumr % absDeno;
        if (rem == 0) return result;

        result += '.';
        unordered_map<int, int> mpp;

        // TC : O(deno + length)
        while (rem) {
            if (mpp.count(rem)) {   // one time
                result.insert(mpp[rem], "(");  // O(length)  
                result += ")";
                break;
            }

            mpp[rem] = result.length();
            
            rem *= 10;
            
            int digit = rem / absDeno;
            result += to_string(digit);

            rem = rem % absDeno;
        }

        return result;
    }
};

File: Leetcode Daily Challenge/027-September-2025/024-fraction-to-recurring-decimal.java
// https://leetcode.com/problems/fraction-to-recurring-decimal/




import java.util.*;

class Solution {
    // TC : O(deno + length)
    // SC : O(deno)
    public String fractionToDecimal(int numerator, int denominator) {
        if (numerator == 0) return "0";

        StringBuilder result = new StringBuilder();

        if ((long)numerator * denominator < 0) {
            result.append("-");
        }

        long absNum = Math.abs((long)numerator);
        long absDen = Math.abs((long)denominator);

        long div = absNum / absDen;
        result.append(div);
        
        long rem = absNum % absDen;
        if (rem == 0) return result.toString();

        result.append(".");
        Map<Long, Integer> mpp = new HashMap<>();

        while (rem > 0) {
            if (mpp.containsKey(rem)) {
                result.insert(mpp.get(rem), "(");
                result.append(")");
                break;
            }

            mpp.put(rem, result.length());

            rem *= 10;
            long digit = rem / absDen;

            result.append(digit);
            rem %= absDen;
        }

        return result.toString();
    }
}